### PR TITLE
PCI vUART Support in HV

### PR DESCRIPTION
**Pull request by taoyuhong**
How to use:
Add pci-vuart in vm_config. Because PCI vUART are also virtual pci devices, their config information are in acrn_vm_pci_dev_config, just like that:

struct acrn_vm_pci_dev_config vm0_pci_devs[] = {
           {
                   .emu_type = PCI_DEV_TYPE_HVEMUL,
                   .vbdf.bits = {.b = 0x00U, .d = 0x04U, .f = 0x00U},
                   .vdev_ops = &vmcs9900_ops,
                   .vbar_base[0] = 0x80001000, /*MMIO bar*/
                   .vbar_base[1] = 0x80002000, /*MSIX bar*/
                   .vuart_idx = 2,
                   .t_vuart.vm_id = 1U,
                   .t_vuart.vuart_id = 1U,
           },
        }

If create for post-launched VM, use '-s,<slot>,uart,vuart_idx:<val>' acrn-dm option

### CODE REVIEW

File: hypervisor/common/hypercall.c [hunk 1]
```diff
@@ -1216,16 +1219,31 @@ int32_t hcall_destroy_vdev(struct acrn_vm *vm, struct acrn_vm *target_vm, __unus
...
+			if (op != NULL) {
+				bdf.value = (uint16_t) dev.slot;
+				vdev = pci_find_vdev(&vm->vpci, bdf);
+				if (vdev != NULL) {
+					vdev->pci_dev_config->vbdf.value = UNASSIGNED_VBDF;
+					ret = 0;
+				} else {
+					pr_warn("%s, failed to destroy emulated device %x:%x.%x\n",
+						__func__, bdf.bits.b, bdf.bits.d, bdf.bits.f);
+					ret = -EINVAL;
```
Comment (dongyaozu) [inline hypervisor/common/hypercall.c:1239]: Why we need this ? It is default value

File: hypervisor/common/hypercall.c [hunk 2]
```diff
@@ -1216,16 +1219,31 @@ int32_t hcall_destroy_vdev(struct acrn_vm *vm, struct acrn_vm *target_vm, __unus
...
+				}
+
+				if (op->destroy != NULL) {
+					ret = op->destroy(vdev);
```
Comment (dongyaozu) [inline hypervisor/common/hypercall.c:1243]: The previous "ret" got in Line 1235 was overwritten then

Comment (dongyaozu) [reply hypervisor/common/hypercall.c:1243]: Or if vdev == NULL, do we need to do this?

File: hypervisor/dm/vpci/vmcs9900.c [hunk 3]
```diff
@@ -167,3 +168,46 @@ const struct pci_vdev_ops vmcs9900_ops = {
...
+	uint32_t i;
+	struct acrn_vm_config *vm_config = get_vm_config(vm->vm_id);
+	struct acrn_vm_pci_dev_config *dev_config = NULL;
+	int32_t ret = -EINVAL;
+	uint16_t vuart_idx = *((uint16_t*)(dev->args));
+
+	for (i = 0U; i < vm_config->pci_dev_num; i++) {
+		dev_config = &vm_config->pci_devs[i];
+		if (dev_config->vuart_idx == vuart_idx) {
+			spinlock_obtain(&vm->vpci.lock);
```
Comment (dongyaozu) [inline hypervisor/dm/vpci/vmcs9900.c:184]: I thought we concluded we didn't need this lock. Not?
@ywan170

Comment (ywan170) [reply hypervisor/dm/vpci/vmcs9900.c:184]: Yes. This lock can be removed.

Comment (taoyuhong) [reply hypervisor/dm/vpci/vmcs9900.c:184]: Sorry， will fix that

File: hypervisor/common/hypercall.c (hunk 2 above)
Comment (taoyuhong) [reply hypervisor/common/hypercall.c:1243]: Yes, that is a bug

File: hypervisor/common/hypercall.c (hunk 1 above)
Comment (taoyuhong) [reply hypervisor/common/hypercall.c:1239]: will remove it

**Review by dongyaozu — APPROVED** [review]

**Pull request merged**